File: gaft/tests/engine_test.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

''' Test case for engine run.
'''

import unittest
from math import sin, cos

from .. import GAEngine
from ..components import BinaryIndividual, DecimalIndividual
from ..components import Population
from ..operators import RouletteWheelSelection
from ..operators import UniformCrossover
from ..operators import FlipBitMutation


class GAEngineTest(unittest.TestCase):

    def setUp(self):
        self.maxDiff = True

    def test_binary_run(self):
        '''
        Make sure GA engine can run correctly.
        '''
        indv_template = BinaryIndividual(ranges=[(0, 10)], eps=0.001)
        population = Population(indv_template=indv_template, size=50).init()

        # Create genetic operators.
        selection = RouletteWheelSelection()
        crossover = UniformCrossover(pc=0.8, pe=0.5)
        mutation = FlipBitMutation(pm=0.1)

        # Create genetic algorithm engine.
        engine = GAEngine(population=population, selection=selection,
                          crossover=crossover, mutation=mutation)

        @engine.fitness_register
        @engine.minimize
        def fitness(indv):
            x, = indv.solution
            return x + 10*sin(5*x) + 7*cos(4*x)

        engine.run(50)

    def test_decimal_run(self):
        '''
        Make sure GA engine can run correctly with decimal encoding individuals.
        '''
        indv_template = DecimalIndividual(ranges=[(0, 10)], eps=0.001)
        population = Population(indv_template=indv_template, size=50).init()

        # Create genetic operators.
        selection = RouletteWheelSelection()
        crossover = UniformCrossover(pc=0.8, pe=0.5)
        mutation = FlipBitMutation(pm=0.1)

        # Create genetic algorithm engine.
        engine = GAEngine(population=population, selection=selection,
                          crossover=crossover, mutation=mutation)

        @engine.fitness_register
        @engine.minimize
        def fitness(indv):
            x, = indv.solution
            return x + 10*sin(5*x) + 7*cos(4*x)

        engine.run(50)

if '__main__' == __name__:
    suite = unittest.TestLoader().loadTestsFromTestCase(GAEngineTest)
    unittest.TextTestRunner(verbosity=2).run(suite)

File: gasol/unittest/engine_test.cpp
/*! \brief Unittest for GA engine.
 */ 

#include "engine.h"
#include "mpiutils.h"
#include "gtest/gtest.h"

#include <cmath>
#include <random>
#include <iostream>


namespace {

double fitness(const gasol::Individual & indv)
{ 
    double x = indv.solution()[0];
    return x + 10*std::sin(5*x) + 7*std::cos(4*x);
}

class EngineTest : public ::testing::Test {

protected:
    virtual void SetUp()
    {
        int size = 50;
        for (int i = 0; i < size; i++)
        {
            gasol::Individual indv(ranges_, precisions_);
            individuals_.push_back(indv);
        }
    }

    virtual void TearDown() {}

    // Individuals used in tests.
    std::vector<std::pair<double, double>> ranges_ {{0.0, 10.0}};
    std::vector<double> precisions_ = {0.001};
    std::vector<gasol::Individual> individuals_;
};

TEST_F(EngineTest, Run)
{
    gasol::Population population(individuals_, &fitness);
    gasol::RouletteWheelSelection selection;
    gasol::UniformCrossover crossover(0.8, 0.5);
    gasol::FlipBitMutation mutation(0.1);

    gasol::Engine engine(population, selection, crossover, mutation);
    gasol::MPIUtils::init();
    engine.run(100);
    gasol::MPIUtils::finalize();

//    std::cout << "best fitness: "
//              << engine.population().fitness()(engine.population().bestIndv())
//              << std::endl;
}

} // namespace

